index.js

```diff
@@ -43,31 +43,35 @@ function cookieParser (secret, options) {
     }
 
     var cookies = req.headers.cookie
-    var secrets = !secret || Array.isArray(secret)
-      ? (secret || [])
-      : [secret]
 
-    req.secret = secrets[0]
-    req.cookies = Object.create(null)
-    req.signedCookies = Object.create(null)
+    function parseCookies(err, secrets) {
+      req.secret = secrets[0]
+      req.cookies = Object.create(null)
+      req.signedCookies = Object.create(null)
 
-    // no cookies
-    if (!cookies) {
-      return next()
-    }
+      // no cookies
+      if (!cookies) {
+        return next()
+      }
 
-    req.cookies = cookie.parse(cookies, options)
+      req.cookies = cookie.parse(cookies, options)
 
-    // parse signed cookies
-    if (secrets.length !== 0) {
-      req.signedCookies = signedCookies(req.cookies, secrets)
-      req.signedCookies = JSONCookies(req.signedCookies)
-    }
+      // parse signed cookies
+      if (secrets.length !== 0) {
+        req.signedCookies = signedCookies(req.cookies, secrets)
+        req.signedCookies = JSONCookies(req.signedCookies)
+      }
 
-    // parse JSON cookies
-    req.cookies = JSONCookies(req.cookies)
+      // parse JSON cookies
+      req.cookies = JSONCookies(req.cookies)
+
+      next()
+    }
 
-    next()
+    if (!secret) parseCookies(null, [])
+    else if (Array.isArray(secret)) parseCookies(null, secret)
+    else if (typeof secret === "function") secret(req, parseCookies)
```

Comment:
the callback function is going to have to also do the same undefined/array conversion.